shadowsocks/config.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"time"
+)
+
+// Config represents a Shadowsocks proxy configuration.
+type Config struct {
+	ID         string
+	Host       string `json:"server"`
+	Port       int    `json:"server_port"`
+	Password   string `json:"password"`
+	Cipher     string `json:"method"`
+	Name       string `json:"remarks,omitempty"`
+	Plugin     string `json:"plugin,omitempty"`
+	PluginOpts string `json:"plugin_opts,omitempty"`
+}
+
+// FetchConfigRequest encapsulates a request to an online config server.
+type FetchConfigRequest struct {
+	// URL is the HTTPs endpoint of an online config server.
+	URL string
+	// Method is the HTTP method to use in the request.
+	Method string
+	// TrustedCertFingerprint is the sha256 hash of the online config server's
+	// TLS certificate.
+	TrustedCertFingerprint []byte
+}
+
+// FetchConfigResponse encapsulates a response from an online config server.
+type FetchConfigResponse struct {
+	// Config is the parsed server response.
+	Config OnlineConfig
```

Comment:
I feel like this is confusing `FetchConfigResponse.Config` is not a `Config`), especially because there's a singular/plural mismatch (it contains a `[]Config`).

Reply:
I further qualified the model: 
- `Config` -> `ProxyConfig`
- `FetchConfig[Request|Response]` -> `OnlineConfig[Request|Response]`
- `FetchConfig` -> `FetchOnlineConfig`
- `FetchConfigResponse.Config` -> `OnlineConfigResponse.OnlineConfig`

Please let me know if this makes sense to you.

Reply:
Another option to address the singular/plural mismatch would be to embed `OnlineConfig` in `OnlineConfigResponse`. That way, `OnlineConfigResponse.Proxies` would be consistent.

Reply:
The latest version looks good to me.